### PR TITLE
fix mismatched tensors when using bidir encoder and attention

### PR DESCRIPTION
**Pull request by alexholdenmiller**

### CODE REVIEW

**Comment by alexholdenmiller** [comment]
fix for #542

File: examples/train_model.py [hunk 1]
```diff
@@ -145,6 +145,8 @@ def validate(self):
         valid_report, self.valid_world = run_eval(
             self.agent, opt, 'valid', opt['validation_max_exs'],
             valid_world=self.valid_world)
+        if opt.get('model_file'):
```
Comment (jaseweston) [inline examples/train_model.py]: what's this for?

Comment (alexholdenmiller) [reply examples/train_model.py]: in case you want to continue training from where you stopped instead of from the best valid so far

Comment (jaseweston) [reply examples/train_model.py]: it won't have self.best_valid in either case will it? or will it?
i would make this an option as it doubles the space on disk

Comment (alexholdenmiller) [reply examples/train_model.py]: we save the best valid result straight to the model_file arg with nothing appended to it. I'll add a flag for last valid though!

File: examples/train_model.py [hunk 2]
```diff
@@ -62,6 +62,9 @@ def setup_args(model_args=None):
                        type=float, default=1.0,
                        help='value at which training will stop if exceeded by '
                             'training metric')
+    train.add_argument('--save-last-valid', type='bool', default=False
+                       help='Saves the model to model_file.last_valid after '
+                            'every validation (overwriting previous saves).')
```
Comment (jaseweston) [inline examples/train_model.py]: needs to explain the best model is still saved in model_file

Comment (jaseweston) [reply examples/train_model.py]: in fact i think something  like  'save-checkpoint'  might be clearer?  then the save-every-n-secs should save there too (and call it   .checkpt or something instead?)

File: parlai/agents/seq2seq/modules.py [hunk 3]
```diff
@@ -117,7 +118,8 @@ def forward(self, xs, ys=None, cands=None, valid_cands=None):
 class Encoder(nn.Module):
     def __init__(self, num_features, padding_idx=0, rnn_class='lstm',
                  emb_size=128, hidden_size=128, num_layers=2, dropout=0.1,
-                 bidirectional=False, shared_lt=None, shared_rnn=None):
+                 bidirectional=False, shared_lt=None, shared_rnn=None,
+                 sparse=False):
```
Comment (emilydinan) [inline parlai/agents/seq2seq/modules.py]: Why do we have this sparse argument here (and in the decoder) when it is always initialized to False?

Comment (alexholdenmiller) [reply parlai/agents/seq2seq/modules.py]: I'd like for it to be an option, but there are some features still missing--this was the first step towards supporting it. maybe it's too soon to commit this section...

**Comment by alexholdenmiller** [comment]
fixed jason's changes, I'd like to merge for now to get this up to date (it does have a bug fix)

**Comment by alexholdenmiller** [comment]
happy to revisit any changes

**Pull request merged**